### ISSUE TITLE
How to overload Parameter

### ISSUE THREAD

**Issue by mzks** [zfit/zfit#284]
I often use interactive fitting for small test on jupyter notebook. I usually draw data and models to check initial parameters.
On zfit, the parameters are defined with initial parameters like
```
mu = zfit.Parameter("mu", 0)
sigma = zfit.Parameter("sigma", 1)
```
When these parameters are not good, I would like to change these parameters.
However, If I change these initial parameters and re-run these statement, zfit will take a message.
https://github.com/zfit/zfit/blob/a798f931cdd3d876d1cdaffd4776781b5d1330b4/zfit/core/parameter.py#L240
I would like to run-then-rerun to find good initial parameters, thus it is better to remove this assert.

My alternative, naive idea is just removing the parameter if it's found
```
class ZfitParameterMixin(BaseNumeric):
    _existing_params = OrderedDict()

    def __init__(self, name, **kwargs):
        if name in self._existing_params:
            del self._existing_params[name]
            #raise NameAlreadyTakenError("Another parameter is already named {}. "
                                        #"Use a different, unique one.".format(name))

        self._existing_params.update({name: self})
        self._name = name
        super().__init__(name=name, **kwargs)
```
Is there some problem?

**Comment by mayou36** [comment]
Thanks for this, it comes up quite a few times indeed. Simply put, the problem is actually in the way the jupyter notebooks work and while it is technically seems easy to resolve, it is not. For example, if you delete a parameter, what should that mean? What about the existing PDFs that have this parameter assigned?

For all points, please have a look at #186 

We're very open for suggestions, but in general, there does not seem to be a consistent solution.

Is it possible to just reset the parameter values to the initial value? Would that work?

**Comment by mzks** [comment]
Thank you very much for explaination.
The `zfit.Parameter()` seems to be placed global.
Instead of overloading parameter, can I reset everything on zfit (without karnel restart)?

**Comment by mayou36** [comment]
No, there is no real way to do that. However, I recommend in general to do a kernel restart.

Alternatively, what can be done is to _reset the values of the parameters_. This can be done via the `Parameter.set_value` method or via `zfit.params.set_values(params, value)` to set in on multiple parameters at once.

Why exactly would you want to reset everything? Do you may have an example?

**Comment by mzks** [comment]
Thank you very much for catching up.

I've created the wrapper to explain how do I use fitting.
https://github.com/mzks/zfitter/blob/main/notebook/zfitter1.ipynb
This is not small example, but it shows what I want to do.
I always fit something with try-and-error and visualization like that.
For future, I would like to switch models.

Basically, jupyter kernel restarting throws away everything, thus it is hard to suggest everyone to do that.
The `zfit.params.set_values()` can set the initial parameter, but I've not find the way to set lower-limit, upper_limit, and step_size.

I understand the overload behavior should not be default, but could you provide the option to overload if the parameter exists?

**Comment by mayou36** [comment]
> but I've not find the way to set lower-limit, upper_limit, and step_size

This can be done parameter by parameter, these are simple attributes, e.g. if you have a list of parameters called `params`, you can do:
```
for param in params:
    param.step_size = ...
	param.lower = ...
	param.upper = ...
```

> I understand the overload behavior should not be default, but could you provide the option to overload if the parameter exists?

Yes sure, you can do that, but as mentioned in #186, it is just unclear what the behavior should be. E.g . what happens with the old models? Do you want to delete them as well? Like delete everything?
If you really want to hack your way around things, what should work is to remove the parameter from `Parameter._existing_params` dict (e.g. with `del Parameter._existing_params[your_name_here]`

This behavior can be achieved by inheriting and catching it in the init (e.g. if the key exists).

However, what I would recommend you in general is to rather write a small function that creates parameters for you: if it exists, it returns the existing ones and set the step size, lower, upper etc to the desired limits. Does that make sense? I would go for this solution

**Comment by mayou36** [comment]
Btw cool project that you have! Funny little package, I like it!

For the drawing, I would recommend you btw to use the mplhep.histplot with the 'errorbar' style, that looks then like HEP plots.

**Comment by mayou36** [comment]
Feel free to close if it is resolved

**Comment by mzks** [comment]
Thank you.

**Comment by dcervenkov** [comment]
@mayou36 I just started playing with zfit and was immediately bit by the unique name handling - "Another parameter is already named...". This reminds me of ROOT and its treating of object names as first-class citizens. By forcing the names to be unique, you are effectively creating global variables. The following code should really work in Python:
```py
def func1():
    x = zfit.Parameter("x", 1.2, -40, 6)
    # do something with x


def func2():
    x = zfit.Parameter("x", 0.5, 0, 1)
    # do something with x

func1()
func2()
```
Why not do [what TensorFlow does](https://www.tensorflow.org/api_docs/python/tf/Variable) and uniquify the names?
